src/client/datascience/jupyter/kernels/kernelExecution.ts

```diff
@@ -32,48 +35,53 @@ const vscodeNotebookEnums = require('vscode') as typeof import('vscode-proposed'
 export class KernelExecution implements IDisposable {
     private readonly cellExecutions = new WeakMap<NotebookCell, CellExecution>();
 
-    private readonly documentExecutions = new WeakMap<NotebookDocument, MultiCancellationTokenSource>();
+    private readonly documentExecutions = new WeakMap<NotebookDocument, CancellationTokenSource>();
+    private readonly stackOfCellsToExecuteByDocument = new WeakMap<NotebookDocument, CellExecution[]>();
 
     private readonly kernelValidated = new WeakMap<NotebookDocument, { kernel: IKernel; promise: Promise<void> }>();
 
     private readonly executionFactory: CellExecutionFactory;
     private readonly disposables: IDisposable[] = [];
+    private readonly pendingExecution = new ChainedExecutions<void>();
+    private readonly pendingCellExecution = new ChainedExecutions<NotebookCellRunState | undefined>();
     private isRawNotebookSupported?: Promise<boolean>;
+    private readonly kernelRestartHandlerAdded = new WeakSet<IKernel>();
+    private _interruptPromise?: Promise<InterruptResult>;
     constructor(
         private readonly kernelProvider: IKernelProvider,
-        private readonly commandManager: ICommandManager,
         errorHandler: IDataScienceErrorHandler,
         editorProvider: INotebookEditorProvider,
         readonly kernelSelectionUsage: IKernelSelectionUsage,
         readonly appShell: IApplicationShell,
         readonly vscNotebook: IVSCodeNotebook,
         readonly metadata: Readonly<KernelConnectionMetadata>,
         private readonly rawNotebookSupported: IRawNotebookSupportedService,
-        context: IExtensionContext
+        context: IExtensionContext,
+        private readonly interruptTimeout: number
     ) {
         this.executionFactory = new CellExecutionFactory(errorHandler, editorProvider, appShell, vscNotebook, context);
     }
 
     @captureTelemetry(Telemetry.ExecuteNativeCell, undefined, true)
     public async executeCell(notebookPromise: Promise<INotebook>, cell: NotebookCell): Promise<void> {
-        // Cannot execute empty cells.
-        if (this.cellExecutions.has(cell) || cell.document.getText().trim().length === 0) {
-            // clear cell output
-            cell.outputs = [];
+        // Return current execution.
+        if (this.cellExecutions.get(cell)) {
+            traceError(`Cell already executing/queued for execution, requested re-execution of ${cell.index}`);
```

Comment:
Is this an error? Just wondering if it should be traceInfo. Seems like if it's requested again we just return the existing result, which seems ok to me. Something that we don't think should ever happen?

Reply:
> Is this an error? 

Kinda yes, error in the work flow. This should never happen. You cannot execute a cell that's already being executed.

> Just wondering if it should be traceInfo. 

I want a traceError so that this stands out, as this should never happen, if it does then we know immediately that something is wrong.
I plan on adding telemetry for this as well. Found that we were re-executing cells that were already executed & that resulted in weird UI state..

> Something that we don't think should ever happen?

Yes, add comments